File: backend-java/src/main/java/com/bffgin/backend/external/ExternalQuery.java
package com.bffgin.backend.external;

import com.bffgin.backend.domain.TaskError;

import java.util.Map;
import java.util.Optional;

/**
 * 外部公開API(/external/v1/tasks)のクエリパラメータ解析だけを担う、副作用の無い純粋な
 * ヘルパー(実サーバーを起動せずに単体テストできるよう、RestErrorMapperと同じ理由で
 * ハンドラ本体から切り出している)
 */
public final class ExternalQuery {

    private ExternalQuery() {
    }

    public record OffsetParams(int page, int pageSize) {
    }

    public record CursorParams(long afterId, int limit) {
    }

    public static long parseUserId(Map<String, String> params) throws TaskError {
        String raw = params.get("user_id");
        if (raw == null || raw.isEmpty()) {
            throw TaskError.userIdRequired();
        }
        try {
            return Long.parseLong(raw);
        } catch (NumberFormatException e) {
            throw TaskError.invalidUserId();
        }
    }

    public static OffsetParams parseOffsetParams(Map<String, String> params) {
        int page = parsePositiveInt(params.get("page"), 1);
        int pageSize = parsePositiveInt(params.get("page_size"), 10);
        return new OffsetParams(page, pageSize);
    }

    public static CursorParams parseCursorParams(Map<String, String> params) {
        String cursorRaw = params.get("cursor");
        long afterId = 0;
        if (cursorRaw != null && !cursorRaw.isEmpty()) {
            try {
                afterId = Long.parseLong(cursorRaw);
            } catch (NumberFormatException ignored) {
                afterId = 0;
            }
        }
        int limit = parsePositiveInt(params.get("limit"), 10);
        return new CursorParams(Math.max(afterId, 0), limit);
    }

    private static int parsePositiveInt(String raw, int defaultValue) {
        if (raw == null || raw.isEmpty()) {
            return defaultValue;
        }
        try {
            int v = Integer.parseInt(raw);
            return Math.max(v, 1);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    public static Optional<String> nextCursor(long lastId) {
        return lastId > 0 ? Optional.of(Long.toString(lastId)) : Optional.empty();
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/rest/ExternalQuery.kt
package com.bffgin.backend.rest

/**
 * 外部公開API(GET /external/v1/tasks)のクエリパラメータ解析・ページング計算。
 * 実サーバー無しで単体テストできるように、純粋関数として切り出す
 * (backend-c/external_query.{h,c}・backend-cppの同等ロジックと同じ分離方針)
 */
object ExternalQuery {

    data class OffsetParams(val page: Int, val pageSize: Int)
    data class CursorParams(val cursor: Long, val limit: Int)

    /** user_idクエリパラメータの必須チェック+数値変換 */
    fun parseUserId(raw: String?): Long {
        if (raw.isNullOrEmpty()) {
            throw ExternalApiError.userIdRequired()
        }
        return raw.toLongOrNull() ?: throw ExternalApiError.invalidUserId()
    }

    /** page既定1・page_size既定10、いずれも1未満は1に補正する(負数・0のpage/page_sizeを弾く) */
    fun parseOffsetParams(pageRaw: String?, pageSizeRaw: String?): OffsetParams {
        val page = (pageRaw?.toIntOrNull() ?: 1).coerceAtLeast(1)
        val pageSize = (pageSizeRaw?.toIntOrNull() ?: 10).coerceAtLeast(1)
        return OffsetParams(page, pageSize)
    }

    /** cursor省略時は0(先頭から)、limit既定10・1未満は1に補正 */
    fun parseCursorParams(cursorRaw: String?, limitRaw: String?): CursorParams {
        val cursor = cursorRaw?.toLongOrNull()?.takeIf { it > 0 } ?: 0L
        val limit = (limitRaw?.toIntOrNull() ?: 10).coerceAtLeast(1)
        return CursorParams(cursor, limit)
    }

    fun offsetToLimitOffset(params: OffsetParams): Pair<Int, Int> {
        val limit = params.pageSize
        val offset = (params.page - 1) * params.pageSize
        return limit to offset
    }
}
